### PR TITLE
Validate checksum of each segment file post download from remote store

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
[Describe what this change achieves]

### Related Issues
- https://github.com/opensearch-project/OpenSearch/issues/8740

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/store/Store.java [hunk 1]
```diff
@@ -952,7 +954,11 @@ public void copyFrom(Directory from, String src, String dest, IOContext context)
             boolean success = false;
             long startTime = System.currentTimeMillis();
             try {
-                super.copyFrom(from, src, dest, context);
+                if (from instanceof RemoteSegmentStoreDirectory) {
+                    copyFileAndValidateChecksum(from, src, dest, context, fileSize);
+                } else {
+                    super.copyFrom(from, src, dest, context);
+                }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/store/Store.java:961]: Why do we need to add an if/else. Also fileSize can be derived from `from` and `src` so can we not implement the `copyFileAndValidateChecksum` to `RemoteSegmentStoreDirectory `

Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/store/Store.java:961]: > Why do we need to add an if/else?

This is a bit defensive to add checksum validation only for remote store downloads. I am not aware of all the flows where file is copied to StoreDirectory and whether it needs a stricter check. It is possible that caller is handling the checksum validation.

> Also fileSize can be derived from from and src so can we not implement the copyFileAndValidateChecksum to RemoteSegmentStoreDirectory 

Here, RemoteSegmentStoreDirectory is `from`. We can't use RemoteSegmentStoreDirectory's `copyFrom` method as in that method `from` would be some another directory.

**Review by gbbafna — APPROVED** [review]

**Review by Bukhtawar — APPROVED** [review]
LGTM

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreRestoreIT.java [hunk 2]
```diff
@@ -10,33 +10,39 @@
...
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import static org.opensearch.node.remotestore.RemoteStoreNodeAttribute.REMOTE_STORE_REPOSITORY_SETTINGS_ATTRIBUTE_KEY_PREFIX;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
 import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertHitCount;
 import static org.hamcrest.Matchers.greaterThan;
 
-@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.SUITE, numDataNodes = 0)
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.TEST, numDataNodes = 0)
```
Comment (andrross) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreRestoreIT.java:45]: What do you think about making a new "RemoteStoreRestoreCorruptionIT" or similar and making that one TEST scope? Test scope will make all the other tests in here more heavyweight.

Comment (sachinpkale) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteStoreRestoreIT.java:45]: Each test in this suite anyway creates its own cluster. So, we were not using benefits of SUITE scope. I changed this to make it more explicit. We need to scan all such remote store test classes that are using TEST scope and change them to SUITE. I will create a tracking issue for the same.

File: server/src/main/java/org/opensearch/index/store/Store.java [hunk 3]
```diff
@@ -984,6 +990,43 @@ public void copyFrom(Directory from, String src, String dest, IOContext context)
...
+                throw new IOException("Exception while reading version info for segment file from remote store: " + dest, e);
+            } finally {
+                if (success == false) {
+                    // If the exception is thrown after file is created, we clean up the file.
+                    // We ignore the exception as the deletion is best-effort basis and can fail if file does not exist.
+                    try {
+                        deleteFile("Quietly deleting", dest);
+                    } catch (Exception e) {
+                        // Ignore
+                    }
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/store/Store.java:1025]: `IOUtils.deleteFilesIgnoringExceptions` takes argument of type `Path`. At this point, we just have a filename and directory implementation keeps base path related information.

File: server/src/main/java/org/opensearch/index/store/Store.java [hunk 4]
```diff
@@ -984,6 +990,43 @@ public void copyFrom(Directory from, String src, String dest, IOContext context)
...
+            throws IOException {
+            RemoteSegmentStoreDirectory.UploadedSegmentMetadata metadata = ((RemoteSegmentStoreDirectory) from)
+                .getSegmentsUploadedToRemoteStore()
+                .get(dest);
+            boolean success = false;
+            try (IndexInput is = from.openInput(src, context); IndexOutput os = createOutput(dest, context)) {
+                // Here, we don't need the exact version as LuceneVerifyingIndexOutput does not verify version
+                // It is just used to emit logs when the entire metadata object is provided as parameter. Also,
+                // we can't provide null version as StoreFileMetadata has non-null check on writtenBy field.
+                Version luceneMajorVersion = Version.parse(metadata.getWrittenByMajor() + ".0.0");
```
[the remark this answers is not in the document]
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/store/Store.java:1003]: @Bukhtawar Do you mean to use like following?

```
Version luceneMajorVersion = Version.parse(metadata.getWrittenByMajor() + "." + V_EMPTY_ID + "." + V_EMPTY_ID);
```

**Pull request merged**